ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/jca/TransactionImporterImple.java

```diff
@@ -122,6 +112,14 @@ public TransactionImple recoverTransaction(Uid actId)
 		if (recovered.baseXid() == null)
 		    throw new IllegalArgumentException();
 		
+		return addImportedTransaction(recovered);
+	}
+
+	private TransactionImple addImportedTransaction(TransactionImple importedTransaction) throws XAException {
+		SubordinateXidImple importedXid = new SubordinateXidImple(importedTransaction.baseXid());
+		TransactionImpleHolder holder = new TransactionImpleHolder();
+		TransactionImpleHolder prevHolder = _transactions.putIfAbsent(importedXid, holder);
+
```

Comment:
Does putIfAbsent do a wait/sleep though? There are some exceptions that could be thrown though: https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentMap.html#putIfAbsent(K,%20V)